alexandria/encoding/src/base64.cairo

```diff
@@ -0,0 +1,246 @@
+use core::clone::Clone;
+use array::ArrayTrait;
+use integer::{upcast, downcast};
+use option::OptionTrait;
+use debug::PrintTrait;
+use alexandria_math::math::{shl, shr, U6_MAX, U8_MAX};
+
+trait Encoder<T> {
+    fn encode(data: Array<T>) -> Array<u8>;
+}
+
+trait Decoder<T> {
+    fn decode(data: Array<T>) -> Array<u8>;
+}
+
+impl Base64Encoder of Encoder<u8> {
+    fn encode(mut data: Array<u8>) -> Array<u8> {
+        let mut char_set = get_base64_char_set();
+        char_set.append('+');
+        char_set.append('/');
+        inner_encode(ref data, ref char_set)
+    }
+}
+
+impl Base64UrlEncoder of Encoder<u8> {
+    fn encode(mut data: Array<u8>) -> Array<u8> {
+        let mut char_set = get_base64_char_set();
+        char_set.append('-');
+        char_set.append('_');
+        inner_encode(ref data, ref char_set)
+    }
+}
+
+impl Base64Decoder of Decoder<u8> {
+    fn decode(mut data: Array<u8>) -> Array<u8> {
+        inner_decode(ref data)
+    }
+}
+
+impl Base64UrlDecoder of Decoder<u8> {
+    fn decode(mut data: Array<u8>) -> Array<u8> {
+        inner_decode(ref data)
+    }
+}
+
+fn inner_encode(ref data: Array<u8>, ref char_set: Array<u8>) -> Array<u8> {
+    let mut p = if (data.len() % 3 == 1) {
+        data.append(0);
+        data.append(0);
+        2
+    } else if (data.len() % 3 == 2) {
+        data.append(0);
+        1
+    } else {
+        0
+    };
+
+    let mut char_set = get_base64_char_set();
+    char_set.append('-');
+    char_set.append('_');
+    let mut result = ArrayTrait::new();
+    encode_loop(p, ref data, 0, ref char_set, ref result);
+    result
+}
+
+fn inner_decode(ref data: Array<u8>) -> Array<u8> {
+    let mut result = ArrayTrait::new();
+    let mut p = 0_u8;
+    if data.len() > 0 {
+        if *data[data.len() - 1] == '=' {
+            p += 1;
+        }
+        if *data[data.len() - 2] == '=' {
+            p += 1;
+        }
+        decode_loop(p, ref data, 0, ref result);
+    }
+    result
+}
+
+fn decode_loop(p: u8, ref data: Array<u8>, d: usize, ref result: Array<u8>) {
```

Comment:
you can use a loop there (maybe you prefer recursive ?)

Reply:
With the current state of Cairo, I prefer recursion